teps/0077-pipeline-partial-execution.md

```diff
@@ -0,0 +1,496 @@
+
+### Goals
+
+- Partially execute a `Pipeline`, by disabling tasks.
+- Resume a failed or cancelled `PipelineRun` from previous `PipelineRun`.
+- Discuss how the results/workspaces of disabled tasks can be referenced or
+  populated from previous `PipelineRun`.
+
+### Non-Goals
+
+- TBD.
+
+### Use Cases (optional)
+
+1. *Optimal use of resources*: `tektoncd` as a backend for ML. 
+   A machine learning pipeline may consist of tasks moving large amount of
+   data and then training ml models, all of it can be very resource consuming
+   and inability to retry would require a user to start the entire pipeline
+   over. A manual retry, with the ability to specify what tasks should
+   be skipped, may be helpful.
+2. Partial execution of pipeline. This is useful for reusing an existing
+   pipeline i.e. a user can disable certain task from an existing pipeline
+   and in this way run it without creating a new pipeline.
+3. Partial execution is also helpful for testing, i.e. skipping some tasks
+   and developing and testing iteratively and quickly.
+4. Pause and resume, i.e. one could manually cancel a running `PipelineRun`
+   and resume at later point.
+
+## Requirements
+
+- Create a new `PipelineRun` to resume or retry a completed `PipelineRun`.
+- Ability to partially run a Pipeline, without editing its definition or without
+  an existing `PipelineRun`.
+- An end user may not have to figure out the execution DAG, to be able to resume
+  or retry.
+
+## Proposal
+
+### Requesting API Changes
+1. Add `pipelineRunRef` under `PipelineRun.spec`. It has following fields: 
+   - `pipelineRunRef.name` which is the name of previously run `PipelineRun`.
+   - `pipelineRunRef.enableTasks` accepts an array of task names under it.
+2. Add `disableTasks` under `PipelineRun.spec`, which accepts an array
+    of task name.
+   - `name`: Name of the task to be disabled.
+
+Q. Why do we need `pipelineRunRef` when we have `disableTasks`? 
+
+`disableTasks` can be used to explicitly disable tasks that a user 
+do not wish to run. On the other hand, in `pipelineRunRef` tekton controller
+automatically figures out the tasks failed and unfinished, because it knows the
+DAG. For the end user, it can be difficult to figure out the DAG and prepare
+the accurate execution plan for the next pipeline run.
+
+Both, `pipelineRunRef` and `disableTasks` are optional fields. See examples
+1 and 2 below.
+
+### Semantics of execution.
+
+- `pipelineRunRef` : pipelineRunRef references a previous pipelineRun and by default
+    selects all the failed and unfinished tasks eligible for retrying/resuming.
+    It references results of completed tasks from previous run.
+
+- `pipelineRunRef.enableTasks`: If a task was successful in previous run, but
+    it is required by the current run, this section can be used to explicitly
+    enable it. For example, a task may perform some initialization for the
+    other tasks in `PipelineRun`.
```

Comment:
This would be on the user to set right ?

Reply:
Yes ! It is not possible to auto-detect it, AFAIK.

Reply:
It might be in future if we add more metadata to our pipelines and workspaces.
We could let users specify that certain tasks are initialisation ones, and need to be executed on re-run.